src/zarr/storage/object_store.py

```diff
@@ -0,0 +1,330 @@
+from __future__ import annotations
+
+import asyncio
+import contextlib
+from collections import defaultdict
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, TypedDict
+
+import obstore as obs
+
+from zarr.abc.store import ByteRangeRequest, Store
+from zarr.core.buffer import Buffer
+from zarr.core.buffer.core import BufferPrototype
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncGenerator, Coroutine, Iterable
+    from typing import Any
+
+    from obstore import ListStream, ObjectMeta, OffsetRange, SuffixRange
+    from obstore.store import ObjectStore as _ObjectStore
+
+    from zarr.core.buffer import Buffer, BufferPrototype
+    from zarr.core.common import BytesLike
+
+
+class ObjectStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, and Azure.
+
+    Parameters
+    ----------
+    store : obstore.store.ObjectStore
+        An obstore store instance that is set up with the proper credentials.
+    """
+
+    store: _ObjectStore
+    """The underlying obstore instance."""
+
+    def __eq__(self, value: object) -> bool:
+        if not isinstance(value, ObjectStore):
+            return False
+
+        return bool(self.store.__eq__(value.store))
+
+    def __init__(self, store: _ObjectStore, *, read_only: bool = False) -> None:
```

Comment:
if `store` is not picklable, you'll need to implement `__getstate__` and `__setstate__` below. The basic strategy you'll want to use is:

- in `__getstate__`, extract the needed config to recreate an the `_ObjectStore`
- in `__setstate__`, use the config from above to recreate the `_ObjectStore` and set `self.store` directly

Reply:
Thanks @jhamman. @kylebarron and I discussed xfailing the pickling tests in this PR and implementing `__getstate__` and `__setstate__`  later on, since `ObjectStore` could be used for non-distributed reading/writing in the meantime. Would you be open to punting this a bit down the road? IIUC Kyle was supportive of adding pickling eventually and I'd be glad to help but it doesn't seem crucial to us right now given that this store will be marked as experimental.

Reply:
Sure, let's add the `__getstate__` methods with a `NotImplementedError` so we can provide a clear error message when things go wrong.